### PR TITLE
mpi4py: enable spawn tests workflow by default

### PR DESCRIPTION
**Pull request by wenduwan**
The previous inter-communicator race condition has been fixed. Enable the workflow by default to catch new regressions.

### CODE REVIEW

File: .github/workflows/ompi_mpi4py.yaml
Comment (jsquyres) [inline .github/workflows/ompi_mpi4py.yaml:134]: Remove the comment, too.

File: .github/workflows/ompi_mpi4py.yaml
Comment (jsquyres) [inline .github/workflows/ompi_mpi4py.yaml:128]: Should the spawn tests just be merged up into the main chunk of tests?  The only reason to run this chunk separately was so that it could fail, and we wouldn't care.  But if the spawn tests are now expected to pass, we might as well run them up in the main tests, and not have a separate section for the spawn tests.

**Review by jsquyres — CHANGES_REQUESTED** [review]

File: .github/workflows/ompi_mpi4py.yaml
Comment (wenduwan) [reply .github/workflows/ompi_mpi4py.yaml:128]: I don't have a strong opinion, though I prefer to keep the test separate for its record in the past.

@hppritcha What do you think?

Comment (hppritcha) [reply .github/workflows/ompi_mpi4py.yaml:128]: Isn't it "cheaper" to do a single github action run?  I vote for merging.

**Review by hppritcha — CHANGES_REQUESTED** [review]
i vote for merging tests into one github action

**Comment by wenduwan** [comment]
@jsquyres @hppritcha Merged workflows as requested.

**Review by hppritcha — APPROVED** [review]

**Review by hppritcha — APPROVED** [review]

**Comment by wenduwan** [comment]
@jsquyres Do you have more comments? Merge is blocked until you approve the change.

**Comment by jsquyres** [comment]
If we're not splitting up into multiple runs on the same build any more, is there any reason to split into 2 jobs (`build` and `run_tests`)?  You could delete the entire upload/download of the build stuff.

**Comment by wenduwan** [comment]
@jsquyres Good point. I merged both `ompi_mpi4py.yaml` and `ompi_mpi4py_tests.yaml` into a single test job. Does it look better?

**Comment by jsquyres** [comment]
Much better, thanks.

Do we really need to test all of these cases?  I ask because the whole thing takes nearly 30 minutes.  Is there value in all of these, or should we cut some of them / make the overall run shorter?

Screenshot 2024-05-31 at 4 25 56 PM

E.g., the final np=5 run with the relocated OMPI -- does that need to be with np=5?  Or do we even need that at all?  I.e., if singleton works with relocated, do we test anything new with np=5?

**Comment by wenduwan** [comment]
Honestly I don't think it's that bad - jenkins is currently the long pole at 50mins, so running mpi4py tests for 30mins sounds benign to me(especially considering bugs that it has revealed in the past).

I'm not sure if np=5 is too big or small... if I were to write the test I might test the edge cases, e.g. singleton, np=3, np=total slots.

@dalcinl What do you think?

**Comment by hppritcha** [comment]
have you watched to see if the np=5 is the slow stage?

**Comment by wenduwan** [comment]
@hppritcha The timing is in https://github.com/open-mpi/ompi/actions/runs/9322808384/job/25664637467?pr=12591

np=5 takes > 4mins.

**Comment by hppritcha** [comment]
i thought the vm where these github actions were running has only 4 "cpus".  Maybe adding ```-bind-to NONE``` to mpiexec cmd line might help?

**Comment by jsquyres** [comment]
Screenshot 2024-05-31 at 5 03 10 PM

**Comment by jsquyres** [comment]
The goal with np=5 might be to intentionally stress the overscribed scenario -- and that's probably a good thing.

But do we need to do that with the relocated OMPI?  I don't think so.  At least for the relocated OMPI, it'll either work or it won't -- I think testing with either singleton or np=1 will tell us if it works, and that's good enough.  Right?

For the other runs -- I think running singleton, np=5, and np=something_else would probably be good.

Do we need all of np=1, np=2, np=3, np=4?  I.e., does running all of the np values tell us something that not running all of them wouldn't tell us?

**Comment by hppritcha** [comment]
I don't think we need np 5 for relocated binary. I agree with @jsquyres aobut that one.
actually it turned out np=3 uncovered some issues with one of the collectives on intercomms or something like that because it had asymmetric local/remote groups.  So lets keep singleton..5 for base testing and Jeff's suggestions for relocated package testing.

**Comment by wenduwan** [comment]
Ack. Removed np=5 relocate test.

**Review by jsquyres — APPROVED** [review]

**Comment by dalcinl** [comment]
@wenduwan I still have to do changes on mpi4py's side for spawn tests to not be skipped under ompi@main. I'm doing a final testing [here](https://github.com/mpi4py/mpi4py-testing/actions/runs/9328735559), and if that goes well, I'll merge mpi4py/mpi4py@64148aeeeb63145e1ac98cd1c92fcb4b8d86e7d3 into master.

Please keep me posted once all these spawn fixes land in v5.0.x, we will have to update things again both in mpi4py and ompi.

**Pull request merged**
